extension/src/time_series/pipeline.rs

```diff
@@ -182,6 +204,302 @@ pub fn lttb_pipeline_element<'p, 'e>(
+                    interval,
+                    downsample_method: downsample_method.into()
+                }
+            }
+        )
+    }
+}
+
+fn downsample(
+    series: &toolkit_experimental::TimeSeries, 
+    element: &toolkit_experimental::Element
+) -> toolkit_experimental::TimeSeries<'static> {
+    let (interval, method) = match element {
+        Element::Downsample{interval, downsample_method} => (interval, downsample_method),
+        _ => panic!("Downsample evaluator called on incorrect pipeline element")
+    };
+    let interval = *interval;
+    let method = DownsampleMethod::from(*method);
+
+    let mut result = None;
+    let mut current = None;
+    let mut points = Vec::new();
+    for point in series.iter() {
+        let TSPoint{ts, ..} = point;
+        let target = (ts + (interval / 2) as i64 - 1) / interval as i64 * interval as i64;
+        if current != Some(target) {
+            if current != None {
+                let new_pt = method.process(&points, current.unwrap(), interval);
+                match result {
+                    None => result = Some(InternalTimeSeries::new_gappy_normal_series(new_pt, interval)),
+                    Some(ref mut series) => series.add_point(new_pt),
+                }
+            }
+            
+            current = Some(target);
+            points = Vec::new();
+        }
+        points.push(point);
+    }
+
+    let new_pt = method.process(&points, current.unwrap(), interval);
+    match result {
+        None => result = Some(InternalTimeSeries::new_gappy_normal_series(new_pt, interval)),
+        Some(ref mut series) => series.add_point(new_pt),
+    }
+
+    TimeSeries::from_internal_time_series(&result.unwrap())
+}
+
+// TODO: there are one or two other gapfill objects in this extension, these should be unified
+pub enum GapfillMethod {
+    LOCF,
+    Interpolate,
+}
+
+impl From<i64> for GapfillMethod {
+    fn from(item: i64) -> Self {
+        match item {
+            1 => GapfillMethod::LOCF,
+            2 => GapfillMethod::Interpolate,
+            _ => panic!("Invalid downsample method")
+        }
+    }
+}
+
+impl Into<i64> for GapfillMethod {
+    fn into(self) -> i64 {
+        match self {
+            GapfillMethod::LOCF => 1,
+            GapfillMethod::Interpolate => 2,
+        }
+    }
+}
+
+impl GapfillMethod {
+    pub fn process<'s>(&self, series: &TimeSeries<'s>) -> MaybeOwnedTs<'s> {
+        unsafe {
+            match series.series {
+                SeriesType::GappyNormalSeries{start_ts, step_interval, count, present, values, ..} => {
+                    match self {
+                        GapfillMethod::LOCF => {
+                            let mut results = Vec::new();
+                            let mut last_val = 0.0;
+                            let mut vidx = 0;
+
+                            for pidx in 0..count {
+                                if present[pidx as usize / 64] & 1 << (pidx % 64) != 0 {
```

Comment:
this should be a method

Reply:
True, but that method belongs on the bitmap type that we need to create for the present array.  I'll save this change until we implement that, as otherwise I'm not really sure where this method should live.